README.rst
.. image:: https://travis-ci.org/python-control/python-control.svg?branch=master
   :target: https://travis-ci.org/python-control/python-control
.. image:: https://coveralls.io/repos/python-control/python-control/badge.png
   :target: https://coveralls.io/r/python-control/python-control

Python Control Systems Library
==============================

The Python Control Systems Library is a Python module that implements basic
operations for analysis and design of feedback control systems.

Features
--------

- Linear input/output systems in state-space and frequency domain
- Block diagram algebra: serial, parallel, and feedback interconnections
- Time response: initial, step, impulse
- Frequency response: Bode and Nyquist plots
- Control analysis: stability, reachability, observability, stability margins
- Control design: eigenvalue placement, linear quadratic regulator
- Estimator design: linear quadratic estimator (Kalman filter)


Links
=====

- Project home page: http://python-control.org
- Source code repository: https://github.com/python-control/python-control
- Documentation: http://python-control.readthedocs.org/
- Issue tracker: https://github.com/python-control/python-control/issues
- Mailing list: http://sourceforge.net/p/python-control/mailman/


Dependencies
============

The package requires numpy, scipy, and matplotlib.  In addition, some routines
use a module called slycot, that is a Python wrapper around some FORTRAN
routines.  Many parts of python-control will work without slycot, but some
functionality is limited or absent, and installation of slycot is recommended
(see below).  Note that in order to install slycot, you will need a FORTRAN
compiler on your machine.  The Slycot wrapper can be found at:

https://github.com/python-control/Slycot

Installation
============

Conda and conda-forge
---------------------

The easiest way to get started with the Control Systems library is
using `Conda <https://conda.io>`_.

The Control Systems library has been packages for the `conda-forge
<https://conda-forge.org>`_ Conda channel, and as of Slycot version
0.3.4, binaries for that package are available for 64-bit Windows,
OSX, and Linux.

To install both the Control Systems library and Slycot in an existing
conda environment, run::

  conda install -c conda-forge control slycot

Pip
---

To install using pip::

  pip install slycot   # optional; see below
  pip install control

If you install Slycot using pip you'll need a development environment
(e.g., Python development files, C and Fortran compilers).

Distutils
---------

To install in your home directory, use::

  python setup.py install --user

To install for all users (on Linux or Mac OS)::

  python setup.py build
  sudo python setup.py install


Development
===========

Code
----

You can check out the latest version of the source code with the command::

  git clone https://github.com/python-control/python-control.git

Testing
-------

You can run the unit tests with `pytest`_ to make sure that everything is
working correctly.  Inside the source directory, run::

  pytest -v

or to test the installed package::

  pytest --pyargs control -v

.. _pytest: https://docs.pytest.org/

License
-------

This is free software released under the terms of `the BSD 3-Clause
License <http://opensource.org/licenses/BSD-3-Clause>`_.  There is no
warranty; not even for merchantability or fitness for a particular
purpose.  Consult LICENSE for copying conditions.

When code is modified or re-distributed, the LICENSE file should
accompany the code or any subset of it, however small.  As an
alternative, the LICENSE text can be copied within files, if so
desired.

Contributing
------------

Your contributions are welcome!  Simply fork the GitHub repository and send a
`pull request`_.

.. _pull request: https://github.com/python-control/python-control/pulls

Please see the `Developer's Wiki`_ for detailed instructions.

.. _Developer's Wiki: https://github.com/python-control/python-control/wiki


ChangeLog
2014-08-09  Richard Murray  <murray@sidamo.local>

	* Cloned python-control/code from SourceForge, mapping SF user names
	to git format (name, e-mail)
	* Pushed to python-control/python-control on GitHub
	* Converted subversion branches/tags to git branches/tags

==== Repository moved from SourceForge to GitHub ====

---- control-0.6d released -----

2014-03-22  Richard Murray  <murray@sidamo.local>

	* src/matlab.py (ss): allow five arguments to create a discrete time
	system.  From Roberto Bucher, 12 Jan 2014

2014-03-22  Richard Murray  <murray@sidamo.local>

	* src/bdalg.py (feedback): allow FRD for feedback()

2014-03-22  Richard Murray  <murray@sidamo.local>

	* tests/discrete_test.py (TestDiscrete.test_sample_system): added
	conversions using tustin and zoh to catch 'dt' bug #5

	* src/dtime.py (sample_system): Changed 'dt' to 'Ts' to fix bug #4

	* tests/frd_test.py (TestFRD.testbdalg): added unit tests for bdalg
	functions operating on FRD data

	* src/bdalg.py (series, parallel, negate, feedback): updated
	documentation to note that FRD ovjects work.
	(feedback): updated conversions to allow for FRD

2013-07-15  Richard Murray  <murray@altura-2.local>

	* src/frdata.py, src/xferfcn.py: updated to use __truediv__ and
	__rtruediv__ for compatibility with python3

2013-07-14  Richard Murray  <murray@altura-2.local>

	* src/dtime.py (sample_system): added check to make sure
	cont2discrete is available in scipy.signal
	(sample_system): added docstring

2013-06-24  Richard Murray  <murray@altura-2.local>

	* tests/minreal_test.py (TestMinreal.testMinrealBrute),
	src/xferfcn.py (_convertToTransferFunction), src/statesp.py
	(StateSpace.__rmul__), src/margins.py (stability_margins): changed
	exception syntax to be python3 compatibile (works for python2.6+)

2013-06-10  Richard Murray  <murray@altura-2.local>

	* src/xferfcn.py (TransferFunction.horner): small fix to docstring

	* src/freqplot.py (nyquist_plot): small fix to docstring

2013-06-09  Richard Murray  <murray@altura-2.local>

	* tests/bdalg_test.py (TestFeedback.testScalarSS)
	(TestFeedback.testScalarTF): added simple test of default argument
	to feedback() for state space and transfer function systems

	* tests/frd_test.py (TestFRD.testFeedback): added test of default
	argument to feedback() for FRD type

	* src/bdalg.py (feedback):
	* src/frdata.py (FRD.feedback):
	* src/statesp.py (StateSpace.feedback):
	patched in default value of 1 for other system in feedback().
	Contributed by Ryan Krauss <rkrauss@siue.edu>

	* tests/xferfcn_test.py (TestXferFcn.testEvalFrMIMO)
	(TestXferFcn.testEvalFrSISO): added tests of __call__ version of
	evalfr.

	* src/xferfcn.py (TransferFunction.__call__): patched in new
	__call__ functionality from Ryan Krauss <rkrauss@siue.edu>, but
	modified to use evalfr (works with MIMO).  SISO transfer functions
	return a single complex number, MIMO transfer functions return a
	matrix of complex numbers.

2013-05-23  Rene van Paassen <rene.vanpaassen@gmail.com>

	* src/margin.py: re-vamped stability_margins function. Now
          uses analytical method for TF and state-space, works on FRD
          objects too, and can optionally return all margins found for
          the analytical method.

	* src/xferfcn.py and src/frdata.py: Now return array result
          for fresp when called with an iterable frequency vector

	* src/xferfcn.py and src/statesp.py: added minreal functions

	* src/bdalg.py: implemented append and connect functions (old
          version, with indexes, not with names)

	* src/matlab.py: added connect, append, minreal

	* src/xferfcn.py and src/statesp.py: added/corrected
          __rmul__. In most cases a gain matrix (np.matrix) may now be
          used in combination with ss or tf systems.

2012-11-10  Richard Murray  <murray@altura.local>

	* src/canonical.py: new module implementing conversions to selected
	canonical forms: canonical_form(), reachable_form() implemented

	* src/lti.py (issiso): new function to quickly check if a system is
	single input, single output or not

2012-11-04  Richard Murray  <murray@altura.local>

	* src/__init__.py: added action item to separate out matlab module
	functionality (eventually)

	* src/matlab.py: added code for setting up MATLAB defaults if matlab
	submodule is imported instead of main (control) module.  See
	comments in __init.py__ for what needs to be done to implement.

	* src/freqplot.py (bode_plot): moved default values of plotting
	options to config module to allow easier use override

	* src/config.py: new module to help set package defaults

	* src/xferfcn.py (_convertToTransferFunction): removed extraneous
	print statements if slycot is not installed

	* src/statesp.py (StateSpace.__str__): fixed error in printing
	timebase when dt is None (python 3)

---- control-0.6c released -----

2012-11-03  Richard Murray  <murray@altura.local>

	* doc/modules.rst: updated documentation format so that items are
	broken down by functionality and not internal module name.  Still
	not complete, but most functions should now show up in documentation.

	* doc/index.rst: got rid of todos from documentation (not formatted
	correctly and shouldn't show up in user documentation)

	* src/lti.py: added missing docstrings

	* doc/intro.rst: small tweaks to text, including adding a link to FBS

2012-11-03  Richard Murray  <murray@altura.local>

	* src/rlocus.py (_RLSortRoots): convert output of range() to
	explicit list for python 3 compatibility

	* tests/modelsimp_test.py, tests/slycot_convert_test.py,
	tests/mateqn_test.py, tests/statefbk_test.py: updated test suites to
	skip specific tests that require slycot if it is not installed

	* src/exception.py (slycot_check): added a function to check if
	slycot is installed

	* README: updated readme to include information about running
	nosetests, which works in python3 (unlike the test_all.py script)

	* tests/test_control_matlab.py, tests/test_all.py: set to executable

control/bdalg.py
"""bdalg.py

This file contains some standard block diagram algebra.

Routines in this module:

append
series
parallel
negate
feedback
connect

"""
"""Copyright (c) 2010 by California Institute of Technology
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.

3. Neither the name of the California Institute of Technology nor
   the names of its contributors may be used to endorse or promote
   products derived from this software without specific prior
   written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL CALTECH
OR THE CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.

Author: Richard M. Murray
Date: 24 May 09
Revised: Kevin K. Chen, Dec 10

$Id$

"""
import numpy as np

from . import frdata as frd
from . import statesp as ss
from . import xferfcn as tf

__all__ = ["series", "parallel", "negate", "feedback", "append", "connect"]


def series(sys1, *sysn):
    """Return the series connection (sysn \\* ... \\*) sys2 \\* sys1

    Parameters
    ----------
    sys1 : scalar, StateSpace, TransferFunction, or FRD
    *sysn : other scalars, StateSpaces, TransferFunctions, or FRDs

    Returns
    -------
    out : scalar, StateSpace, or TransferFunction

    Raises
    ------
    ValueError
        if `sys2.ninputs` does not equal `sys1.noutputs`
        if `sys1.dt` is not compatible with `sys2.dt`

    See Also
    --------
    parallel
    feedback

    Notes
    -----
    This function is a wrapper for the __mul__ function in the StateSpace and
    TransferFunction classes.  The output type is usually the type of `sys2`.
    If `sys2` is a scalar, then the output type is the type of `sys1`.

    If both systems have a defined timebase (dt = 0 for continuous time,
    dt > 0 for discrete time), then the timebase for both systems must
    match.  If only one of the system has a timebase, the return
    timebase will be set to match it.

    Examples
    --------
    >>> sys3 = series(sys1, sys2) # Same as sys3 = sys2 * sys1

    >>> sys5 = series(sys1, sys2, sys3, sys4) # More systems

    """
    from functools import reduce

    return reduce(lambda x, y: y * x, sysn, sys1)


def parallel(sys1, *sysn):
    """
    Return the parallel connection sys1 + sys2 (+ ... + sysn)

    Parameters
    ----------
    sys1 : scalar, StateSpace, TransferFunction, or FRD
    *sysn : other scalars, StateSpaces, TransferFunctions, or FRDs

    Returns
    -------
    out : scalar, StateSpace, or TransferFunction

    Raises
    ------
    ValueError
        if `sys1` and `sys2` do not have the same numbers of inputs and outputs

    See Also
    --------
    series
    feedback

    Notes
    -----
    This function is a wrapper for the __add__ function in the
    StateSpace and TransferFunction classes.  The output type is usually
    the type of `sys1`.  If `sys1` is a scalar, then the output type is
    the type of `sys2`.

    If both systems have a defined timebase (dt = 0 for continuous time,
    dt > 0 for discrete time), then the timebase for both systems must
    match.  If only one of the system has a timebase, the return
    timebase will be set to match it.

    Examples
    --------
    >>> sys3 = parallel(sys1, sys2) # Same as sys3 = sys1 + sys2

    >>> sys5 = parallel(sys1, sys2, sys3, sys4) # More systems

    """
    from functools import reduce

    return reduce(lambda x, y: x + y, sysn, sys1)


def negate(sys):
    """
    Return the negative of a system.

    Parameters
    ----------
    sys : StateSpace, TransferFunction or FRD

    Returns
    -------
    out : StateSpace or TransferFunction

    Notes
    -----
    This function is a wrapper for the __neg__ function in the StateSpace and
    TransferFunction classes.  The output type is the same as the input type.

    Examples
    --------
    >>> sys2 = negate(sys1) # Same as sys2 = -sys1.

    """
    return -sys


#! TODO: expand to allow sys2 default to work in MIMO case?
def feedback(sys1, sys2=1, sign=-1):
    """
    Feedback interconnection between two I/O systems.

    Parameters
    ----------
    sys1 : scalar, StateSpace, TransferFunction, FRD
        The primary process.
    sys2 : scalar, StateSpace, TransferFunction, FRD
        The feedback process (often a feedback controller).
    sign: scalar
        The sign of feedback.  `sign` = -1 indicates negative feedback, and
        `sign` = 1 indicates positive feedback.  `sign` is an optional
        argument; it assumes a value of -1 if not specified.

    Returns
    -------
    out : StateSpace or TransferFunction

    Raises
    ------
    ValueError
        if `sys1` does not have as many inputs as `sys2` has outputs, or if
        `sys2` does not have as many inputs as `sys1` has outputs
    NotImplementedError
        if an attempt is made to perform a feedback on a MIMO TransferFunction
        object

    See Also
    --------
    series
    parallel

    Notes
    -----
    This function is a wrapper for the feedback function in the StateSpace and
    TransferFunction classes.  It calls TransferFunction.feedback if `sys1` is a
    TransferFunction object, and StateSpace.feedback if `sys1` is a StateSpace
    object.  If `sys1` is a scalar, then it is converted to `sys2`'s type, and
    the corresponding feedback function is used.  If `sys1` and `sys2` are both
    scalars, then TransferFunction.feedback is used.

    """
    # Allow anything with a feedback function to call that function
    try:
        return sys1.feedback(sys2, sign)
    except AttributeError:
        pass

    # Check for correct input types.
    if not isinstance(
        sys1,
        (int, float, complex, np.number, tf.TransferFunction, ss.StateSpace, frd.FRD),
    ):
        raise TypeError(
            "sys1 must be a TransferFunction, StateSpace "
            + "or FRD object, or a scalar."
        )
    if not isinstance(
        sys2,
        (int, float, complex, np.number, tf.TransferFunction, ss.StateSpace, frd.FRD),
    ):
        raise TypeError(
            "sys2 must be a TransferFunction, StateSpace "
            + "or FRD object, or a scalar."
        )

    # If sys1 is a scalar, convert it to the appropriate LTI type so that we can
    # its feedback member function.
    if isinstance(sys1, (int, float, complex, np.number)):
        if isinstance(sys2, tf.TransferFunction):
            sys1 = tf._convert_to_transfer_function(sys1)
        elif isinstance(sys2, ss.StateSpace):
            sys1 = ss._convert_to_statespace(sys1)
        elif isinstance(sys2, frd.FRD):
            sys1 = frd._convert_to_FRD(sys1, sys2.omega)
        else:  # sys2 is a scalar.
            sys1 = tf._convert_to_transfer_function(sys1)
            sys2 = tf._convert_to_transfer_function(sys2)

    return sys1.feedback(sys2, sign)


def append(*sys):
    """append(sys1, sys2, ..., sysn)

    Group models by appending their inputs and outputs

    Forms an augmented system model, and appends the inputs and
    outputs together. The system type will be the type of the first
    system given; if you mix state-space systems and gain matrices,
    make sure the gain matrices are not first.

    Parameters
    ----------
    sys1, sys2, ..., sysn: StateSpace or Transferfunction
        LTI systems to combine


    Returns
    -------
    sys: LTI system
        Combined LTI system, with input/output vectors consisting of all
        input/output vectors appended

    Examples
    --------
    >>> sys1 = ss([[1., -2], [3., -4]], [[5.], [7]]", [[6., 8]], [[9.]])
    >>> sys2 = ss([[-1.]], [[1.]], [[1.]], [[0.]])
    >>> sys = append(sys1, sys2)

    """
    s1 = ss._convert_to_statespace(sys[0])
    for s in sys[1:]:
        s1 = s1.append(s)
    return s1


def connect(sys, Q, inputv, outputv):
    """Index-based interconnection of an LTI system.

    The system `sys` is a system typically constructed with `append`, with
    multiple inputs and outputs.  The inputs and outputs are connected
    according to the interconnection matrix `Q`, and then the final inputs and
    outputs are trimmed according to the inputs and outputs listed in `inputv`
    and `outputv`.

    NOTE: Inputs and outputs are indexed starting at 1 and negative values
    correspond to a negative feedback interconnection.

    Parameters
    ----------
    sys : StateSpace Transferfunction
        System to be connected
    Q : 2D array
        Interconnection matrix. First column gives the input to be connected.
        The second column gives the index of an output that is to be fed into
        that input. Each additional column gives the index of an additional
        input that may be optionally added to that input. Negative
        values mean the feedback is negative. A zero value is ignored. Inputs
        and outputs are indexed starting at 1 to communicate sign information.
    inputv : 1D array
        list of final external inputs, indexed starting at 1
    outputv : 1D array
        list of final external outputs, indexed starting at 1

    Returns
    -------
    sys: LTI system
        Connected and trimmed LTI system

    Examples
    --------
    >>> sys1 = ss([[1., -2], [3., -4]], [[5.], [7]], [[6, 8]], [[9.]])
    >>> sys2 = ss([[-1.]], [[1.]], [[1.]], [[0.]])
    >>> sys = append(sys1, sys2)
    >>> Q = [[1, 2], [2, -1]]  # negative feedback interconnection
    >>> sysc = connect(sys, Q, [2], [1, 2])

    Notes
    -----
    The :func:`~control.interconnect` function in the
    :ref:`input/output systems <iosys-module>` module allows the use
    of named signals and provides an alternative method for
    interconnecting multiple systems.

    """
    inputv, outputv, Q = np.atleast_1d(inputv), np.atleast_1d(outputv), np.atleast_1d(Q)
    # check indices
    index_errors = (inputv - 1 > sys.ninputs) | (inputv < 1)
    if np.any(index_errors):
        raise IndexError(
            "inputv index %s out of bounds" % inputv[np.where(index_errors)]
        )
    index_errors = (outputv - 1 > sys.noutputs) | (outputv < 1)
    if np.any(index_errors):
        raise IndexError(
            "outputv index %s out of bounds" % outputv[np.where(index_errors)]
        )
    index_errors = (Q[:, 0:1] - 1 > sys.ninputs) | (Q[:, 0:1] < 1)
    if np.any(index_errors):
        raise IndexError("Q input index %s out of bounds" % Q[np.where(index_errors)])
    index_errors = np.abs(Q[:, 1:]) - 1 > sys.noutputs
    if np.any(index_errors):
        raise IndexError("Q output index %s out of bounds" % Q[np.where(index_errors)])

    # first connect
    K = np.zeros((sys.ninputs, sys.noutputs))
    for r in np.array(Q).astype(int):
        inp = r[0] - 1
        for outp in r[1:]:
            if outp < 0:
                K[inp, -outp - 1] = -1.0
            elif outp > 0:
                K[inp, outp - 1] = 1.0
    sys = sys.feedback(np.array(K), sign=1)

    # now trim
    Ytrim = np.zeros((len(outputv), sys.noutputs))
    Utrim = np.zeros((sys.ninputs, len(inputv)))
    for i, u in enumerate(inputv):
        Utrim[u - 1, i] = 1.0
    for i, y in enumerate(outputv):
        Ytrim[i, y - 1] = 1.0

    return Ytrim * sys * Utrim
